drv/ignition-api/src/lib.rs

```diff
@@ -33,105 +36,395 @@ pub const PORT_MAX: usize = 40;
+        self.raw_port_state(port)
+            .map(|state| Port::from(state).target)
+    }
+
+    /// Send the given system power `Request` to the given port. Once a request
+    /// is sent and accepted by the Target, it enforces a cooldown before
+    /// subsequent requests are accepted and processed. `SystemPowerOff` and
+    /// `SystemPowerOn` requests have a three second cooldown while a
+    /// `SystemPowerReset` has a six seconds cooldown.
+    #[inline]
+    pub fn send_request(
+        &self,
+        port: u8,
+        request: Request,
+    ) -> Result<(), IgnitionError> {
+        self.controller.send_request(port, request)
+    }
+
+    /// Return the `Counters` for the given port. This function has the
+    /// side-effect of clearing the counters.
+    #[inline]
+    pub fn counters(&self, port: u8) -> Result<Counters, IgnitionError> {
+        self.controller.counters(port)
+    }
+
+    /// Return the `LinkEvents` for the given transceiver and port. See
+    /// `TransceiverSelect` for more details.
+    pub fn link_events(
+        &self,
+        port: u8,
+        txr: TransceiverSelect,
+    ) -> Result<LinkEvents, IgnitionError> {
+        self.controller.link_events(port, txr).map(LinkEvents::from)
+    }
+
+    /// Clear the link events for the given transceiver and port. See
+    /// `TransceiverSelect` for more details.
+    pub fn clear_link_events(
+        &self,
+        port: u8,
+        txr: TransceiverSelect,
+    ) -> Result<(), IgnitionError> {
+        self.controller.clear_link_events(port, txr)
+    }
+
+    /// Return the state of all ports in a single operation. Be aware that this
+    /// reply is fairly large and may require enlarging the stack of the caller.
+    pub fn all_ports(
+        &self,
+    ) -> Result<[Port; PORT_MAX as usize], IgnitionError> {
+        Ok(self.controller.all_port_state()?.map(Port::from))
+    }
+
+    /// Return the `LinkEvents` for all ports and transceivers in a single
+    /// operation.
+    pub fn all_link_events(
+        &self,
+    ) -> Result<[[LinkEvents; 3]; PORT_MAX as usize], IgnitionError> {
+        let data = self.controller.all_link_events()?;
+        let mut link_events = [[Default::default(); 3]; PORT_MAX as usize];
+
+        for port in 0..PORT_MAX as usize {
+            for txr in 0..TransceiverSelect::ALL.len() {
+                link_events[port][txr] = LinkEvents::from(data[port][txr]);
+            }
         }
+
+        Ok(link_events)
+    }
+}
+
+/// `PortState` is an opague type representing (most of) the state of an
+/// Ignition Controller port. It is highly dependent on the RTL implementation
+/// of the system and the use of the `Port` and `Target` types is encouraged
+/// instead.
+#[derive(
+    Copy,
+    Clone,
+    Debug,
+    Default,
+    PartialEq,
+    Eq,
+    FromPrimitive,
+    From,
+    FromBytes,
+    AsBytes,
+)]
+#[repr(C)]
+pub struct PortState(u64);
```

Comment:
It may have to be public to be part of the idol interface?

Reply:
Yes, I do want to use it in the Idol interface (because I want the `-server` to serve up a New Type rather than just a u64 and enforce that in the `From` implementations). The compromise here is that while visible it's opaque to the rest of the system.